File: src/main/scala/com/github/mauricio/async/db/postgresql/parsers/InformationParser.scala
package com.github.mauricio.postgresql.parsers

import com.github.mauricio.postgresql.ChannelUtils
import com.github.mauricio.postgresql.messages.backend.Message
import java.nio.charset.Charset
import org.jboss.netty.buffer.ChannelBuffer

/**
 * User: mauricio
 * Date: 3/31/13
 * Time: 12:54 AM
 */
abstract class InformationParser( charset : Charset ) extends Decoder {

  override def parseMessage(b: ChannelBuffer): Message = {

    val fields = scala.collection.mutable.Map[Char,String]()

    while ( b.readable() ) {
      val kind = b.readByte()

      if ( kind != 0 ) {
        fields.put(
          kind.toChar,
          ChannelUtils.readCString(b, charset)
        )
      }

    }

    createMessage(fields.toMap)
  }

  def createMessage( fields : Map[Char,String] ) : Message

}

File: postgresql-async/src/main/java/com/github/jasync/sql/db/postgresql/parsers/InformationParser.kt
package com.github.jasync.sql.db.postgresql.parsers

import com.github.jasync.sql.db.postgresql.messages.backend.ServerMessage
import com.github.jasync.sql.db.util.ByteBufferUtils
import io.netty.buffer.ByteBuf
import java.nio.charset.Charset

abstract class InformationParser(val charset: Charset) : MessageParser {

    override fun parseMessage(buffer: ByteBuf): ServerMessage {

        val fields = mutableMapOf<Char, String>()
        while (buffer.isReadable) {
            val kind = buffer.readByte()
            if (kind.toInt() != 0) {
                fields[kind.toChar()] = ByteBufferUtils.readCString(buffer, charset)
            }
        }

        return createMessage(fields.toMap())
    }

    abstract fun createMessage(fields: Map<Char, String>): ServerMessage

}
